### PR TITLE
Fix directory bucket region name retrieval

### PR DESCRIPTION
**Pull request by kriscon-db**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/kriscon-db/mlflow/pull/10967?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10967/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10967
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Updated the get region logic for the _get_region_name method to handle the non-error response from the boto client in optimized_s3_artifact_repo.py

### How is this PR tested?

- [ x] Existing unit/integration tests
- [x ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/store/artifact/optimized_s3_artifact_repo.py [hunk 1]
```diff
@@ -65,9 +70,27 @@ def _get_region_name(self):
...
-            return temp_client.head_bucket(Bucket=self.bucket)["BucketRegion"]
+            head_bucket_resp = temp_client.head_bucket(Bucket=self.bucket)
+            if _BUCKET_REGION in head_bucket_resp:
+                return head_bucket_resp[_BUCKET_REGION]
+            if (
+                _RESPONSE_METADATA in head_bucket_resp
+                and _HTTP_HEADERS in head_bucket_resp[_RESPONSE_METADATA]
+                and _HTTP_HEADER_BUCKET_REGION
+                in head_bucket_resp[_RESPONSE_METADATA][_HTTP_HEADERS]
+            ):
```
Comment (prithvikannan) [inline mlflow/store/artifact/optimized_s3_artifact_repo.py:81]: nit: can we leave a comment explaining what these various branches represent?

File: mlflow/store/artifact/optimized_s3_artifact_repo.py [hunk 2]
```diff
@@ -65,9 +70,27 @@ def _get_region_name(self):
...
+                ]
+            if _BUCKET_LOCATION_NAME in head_bucket_resp:
+                _logger.warning(
+                    f"Directory bucket {self.bucket} found with BucketLocationName "
+                    f"{head_bucket_resp[_BUCKET_LOCATION_NAME]}."
+                )
+            raise Exception(f"Unable to get the region name for bucket {self.bucket}.")
         except ClientError as error:
-            return error.response["ResponseMetadata"]["HTTPHeaders"]["x-amz-bucket-region"]
+            _logger.info("Client error caught and getting region from the headers.")
```
Comment (prithvikannan) [inline mlflow/store/artifact/optimized_s3_artifact_repo.py:92]: dont fully understand this block, but should this be logger.warn or logger.error instead?

**Review by prithvikannan — APPROVED** [review]
LGTM, thanks @kriscon-db for the quick turnaround

File: mlflow/store/artifact/optimized_s3_artifact_repo.py (hunk 2 above)
Comment (kriscon-db) [reply mlflow/store/artifact/optimized_s3_artifact_repo.py:92]: It's not really an error.  Also, I should remove this....this was part of the extra logging for validation that I had added.

File: mlflow/store/artifact/optimized_s3_artifact_repo.py (hunk 1 above)
Comment (kriscon-db) [reply mlflow/store/artifact/optimized_s3_artifact_repo.py:81]: sure.

File: mlflow/store/artifact/optimized_s3_artifact_repo.py [hunk 3]
```diff
@@ -65,9 +70,38 @@ def _get_region_name(self):
...
+                    f"Directory bucket {self.bucket} found with BucketLocationName "
+                    f"{head_bucket_resp[_BUCKET_LOCATION_NAME]}."
+                )
+            raise Exception(f"Unable to get the region name for bucket {self.bucket}.")
         except ClientError as error:
-            return error.response["ResponseMetadata"]["HTTPHeaders"]["x-amz-bucket-region"]
+            # If a client error occurs, we check to see if the x-amz-bucket-region field is set
+            # in the response and return that.  If it is not present, this will raise due to the
+            # key not being present.
+            return error.response[_RESPONSE_METADATA][_HTTP_HEADERS][_HTTP_HEADER_BUCKET_REGION]
```
Comment (kriscon-db) [inline mlflow/store/artifact/optimized_s3_artifact_repo.py:104]: I guess the only question I have is whether we want to test for the key's presence and if it is not there, raise a more user friendly exception?  cc @smurching on this question

**Comment by kriscon-db** [comment]
@mlflow-automation autoformat

**Comment by kriscon-db** [comment]
@mlflow-automation autoformat

**Pull request merged**